code/modules/lighting/lighting_mask/shadow_calculator.dm

```diff
@@ -0,0 +1,529 @@
+//#define SHADOW_DEBUG
+//Lighting texture scales in world units (divide by 32)
+//256 = 8,4,2
+//1024 = 32,16,8
+#define LIGHTING_SHADOW_TEX_SIZE 8
+
+#define COORD_LIST_ADD(listtoadd, x, y) \
+	if(islist(listtoadd["[x]"])) { \
+		var/list/_L = listtoadd["[x]"]; \
+		BINARY_INSERT_NUM(y, _L); \
+	} else { \
+		listtoadd["[x]"] = list(y);\
+	}
+
+#ifdef SHADOW_DEBUG
+#define DEBUG_HIGHLIGHT(x, y, colour) \
+	do { \
+		var/turf/T = locate(x, y, 2); \
+		if(T) { \
+			T.color = colour; \
+		}\
+	} while (0)
+#define DO_SOMETHING_IF_DEBUGGING_SHADOWS(something) something
+#else
+#define DEBUG_HIGHLIGHT(x, y, colour)
+#define DO_SOMETHING_IF_DEBUGGING_SHADOWS(something)
+#endif
+
+/atom/movable/lighting_mask
+	var/list/affecting_turfs
+	var/list/mutable_appearance/shadows
+	var/times_calculated = 0
+
+/atom/movable/lighting_mask/Destroy()
+	//Make sure we werent destroyed in init
+	if(!SSlighting.started)
+		SSlighting.sources_that_need_updating -= src
+	//Remove from affecting turfs
+	if(affecting_turfs)
+		for(var/turf/thing as() in affecting_turfs)
+			LAZYREMOVE(thing.lights_affecting, src)
+		affecting_turfs = null
+	//Cut the shadows. Since they are overlays they will be deleted when cut from overlays probably.
+	LAZYCLEARLIST(shadows)
+	. = ..()
+
+/atom/movable/lighting_mask/proc/link_turf_to_light(turf/T)
+	LAZYOR(affecting_turfs, T)
+	LAZYOR(T.lights_affecting, src)
+
+/atom/movable/lighting_mask/proc/unlink_turf_from_light(turf/T)
+	LAZYREMOVE(affecting_turfs, T)
+	LAZYREMOVE(T.lights_affecting, src)
+
+//Returns a list of matrices corresponding to the matrices that should be applied to triangles of
+//coordinates (0,0),(1,0),(0,1) to create a triangcalculate_shadows_matricesle that respresents the shadows
+//takes in the old turf to smoothly animate shadow movement
+/atom/movable/lighting_mask/proc/calculate_lighting_shadows()
+
+	//Check to make sure lighting is actually started
+	//If not count the amount of duplicate requests created.
+	if(!SSlighting.started)
+		if(awaiting_update)
+			SSlighting.duplicate_shadow_updates_in_init ++
+			return
+		SSlighting.sources_that_need_updating += src
+		awaiting_update = TRUE
+		return
+
+	//Dont bother calculating at all for small shadows
+	var/range = radius * 0.5
+
+	if(range < 1)
+		return
+
+	SSlighting.total_shadow_calculations ++
+
+	var/unrounded_range = range
+	range = FLOOR(unrounded_range, 1)
+	if(unrounded_range > range)
+		range ++
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(var/timer = TICK_USAGE)
+
+	//Remove the old shadows
+	overlays.Cut()
+
+	//Optimise grouping by storing as
+	// Key : x (AS A STRING BECAUSE BYOND DOESNT ALLOW FOR INT KEY DICTIONARIES)
+	// Value: List(y values)
+	var/list/opaque_atoms_in_view = list()
+
+	//Reset the list
+	if(islist(affecting_turfs))
+		for(var/turf/T in affecting_turfs)
+			T?.lights_affecting -= src
```

Comment:
if your're going to do this, `?.` isn't needed - that'd only be needed with `as()`

Reply:
Ah, should definitely be an as() list